Clean-up error checking in linalg.

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80767

This PR does a number of things:
- Fixes https://github.com/pytorch/pytorch/issues/80735 returning early in the most common case (no errors).
- Leaves just one entry to check the errors, so that it's as uniform as it gets (easy to get right, difficult to get wrong)
- Having all the error checking in one function simplifies the code
- We make sure that the `infos_ptr` sent over to LAPACK is always of type `int`, as it should
- We just access `infos.data_ptr<int32_t>` in `_linalg_check_errors`

Edit. I get a consistent x7-10 speed-up with this patch using the script in https://github.com/pytorch/pytorch/issues/80735, so I think we're good here.

lezcano committed:
Clean-up error checking in linalg.

This PR does a number of things:
- Fixes https://github.com/pytorch/pytorch/issues/80735 returning early in the most common case (no errors).
- Leaves just one entry to check the errors, so that it's as uniform as it gets (easy to get right, difficult to get wrong)
- Having all the error checking in one function simplifies the code
- We make sure that the `infos_ptr` sent over to LAPACK is always of type `int`, as it should
- We just access `infos.data_ptr<int32_t>` in `_linalg_check_errors`

Fixes https://github.com/pytorch/pytorch/issues/80735

[ghstack-poisoned]

lezcano committed:
Update on "Clean-up error checking in linalg."


This PR does a number of things:
- Fixes https://github.com/pytorch/pytorch/issues/80735 returning early in the most common case (no errors).
- Leaves just one entry to check the errors, so that it's as uniform as it gets (easy to get right, difficult to get wrong)
- Having all the error checking in one function simplifies the code
- We make sure that the `infos_ptr` sent over to LAPACK is always of type `int`, as it should
- We just access `infos.data_ptr<int32_t>` in `_linalg_check_errors`

Edit. I get a consistent x7-10 speed-up with this patch using the script in https://github.com/pytorch/pytorch/issues/80735, so I think we're good here.

[ghstack-poisoned]

Lezcano commented:
@IvanYashchuk PTAL

IvanYashchuk commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -299,83 +299,6 @@ static inline void checkFloatingOrComplex(const Tensor& t, const char* const f_n
               f_name, ": Expected a floating point or complex tensor as input. Got ", toString(t.scalar_type()));
 }
 
-/*
```
What's the benefit of moving the code from here to `aten/src/ATen/native/BatchLinearAlgebra.cpp` instead of introducing the changes in this file?

Lezcano replied:
Now there's just one entry point to this function (through `_linalg_check_errors`) so that way it's impossible to use wrong and we always get something that's composite compliant and all that.

Lezcano commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp:
```diff
@@ -2355,7 +2353,7 @@ REGISTER_CUDA_DISPATCH(linalg_eigh_stub, &linalg_eigh_kernel);
 
 template <typename scalar_t>
 static void apply_eig(const Tensor& self, bool eigenvectors, Tensor& out_eigvals, Tensor& out_eigvecs,
```
The day we're able to merge those PRs and remove all that legacy code I'll ship a bottle of champagne to Helsinki.

lezcano committed:
Update on "Clean-up error checking in linalg."


This PR does a number of things:
- Fixes https://github.com/pytorch/pytorch/issues/80735 returning early in the most common case (no errors).
- Leaves just one entry to check the errors, so that it's as uniform as it gets (easy to get right, difficult to get wrong)
- Having all the error checking in one function simplifies the code
- We make sure that the `infos_ptr` sent over to LAPACK is always of type `int`, as it should
- We just access `infos.data_ptr<int32_t>` in `_linalg_check_errors`

Edit. I get a consistent x7-10 speed-up with this patch using the script in https://github.com/pytorch/pytorch/issues/80735, so I think we're good here.

[ghstack-poisoned]

Lezcano commented:
@IvanYashchuk addressed the review PTAL.

lezcano committed:
Update on "Clean-up error checking in linalg."


This PR does a number of things:
- Fixes https://github.com/pytorch/pytorch/issues/80735 returning early in the most common case (no errors).
- Leaves just one entry to check the errors, so that it's as uniform as it gets (easy to get right, difficult to get wrong)
- Having all the error checking in one function simplifies the code
- We make sure that the `infos_ptr` sent over to LAPACK is always of type `int`, as it should
- We just access `infos.data_ptr<int32_t>` in `_linalg_check_errors`

Edit. I get a consistent x7-10 speed-up with this patch using the script in https://github.com/pytorch/pytorch/issues/80735, so I think we're good here.

[ghstack-poisoned]

ngimel commented:
@IvanYashchuk is this PR good to go?

IvanYashchuk reviewed (approved):
Good to go. Thank you, Mario!

IvanYashchuk commented:
@pytorchbot merge -g

Lezcano commented:
@pytorchbot rebase

lezcano committed:
Update on "Clean-up error checking in linalg."


This PR does a number of things:
- Fixes https://github.com/pytorch/pytorch/issues/80735 returning early in the most common case (no errors).
- Leaves just one entry to check the errors, so that it's as uniform as it gets (easy to get right, difficult to get wrong)
- Having all the error checking in one function simplifies the code
- We make sure that the `infos_ptr` sent over to LAPACK is always of type `int`, as it should
- We just access `infos.data_ptr<int32_t>` in `_linalg_check_errors`

Edit. I get a consistent x7-10 speed-up with this patch using the script in https://github.com/pytorch/pytorch/issues/80735, so I think we're good here.

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge -g

Lezcano commented:
@janeyx99 I believe that I should be able to merge this after Ivan approved it, but the merge has failed. Could you check what's going on there?

ngimel commented:
@pytorchbot merge
